source/vir/src/modes.rs

```diff
@@ -98,7 +177,12 @@ fn add_pattern(typing: &mut Typing, mode: Mode, pattern: &Pattern) -> Result<(),
     }
 }
 
-fn check_expr(typing: &mut Typing, outer_mode: Mode, expr: &Expr) -> Result<Mode, VirErr> {
+fn check_expr(
```

Comment:
I think it would be cleaner to make atomic_insts a field of Typing rather than an additional parameter to check_expr and check_stmt.